File: blind75/linkedList/19-removeNthNodeFromEndOfList.js
// 19. Remove Nth Node From End of List
// Given the head of a linked list, remove the nth node from the end of the list and return its head.

// Definition for singly-linked list.
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let dummy = new ListNode();
    dummy.next = head;
    let fast = head;
    let slow = dummy;
    // fast need n node before slow
    while (n > 0) {
        fast = fast.next;
        n = n - 1;
    }

    // when fast arrive in the end 
    while(fast !== null) {
        fast = fast.next;
        slow = slow.next;
    }

    let tmp = slow.next.next;
    slow.next = tmp;

    return dummy.next;
};

// 这道题的思路是用两个指针，一个在前，一个在后
// 前面的那个指针提前n个node后指针
// 前面的到达终点的时候，
// 后面的那个到达了len - n
// 需要用一个dummy来到达 len-n node 的前面一个

let head = [1,2,3,4,5], n = 2;
let head1 = [1], n1 = 1;
let head2 = [1,2], n2 = 1;
console.log(removeNthFromEnd(head, n));
console.log(removeNthFromEnd(head1, n1));
console.log(removeNthFromEnd(head2, n2));

File: specified-learning专项训练/linkedList/19-removeNthNodeFromEndOfList.py
# 19. Remove Nth Node From End of List
# Given the head of a linked list, remove the nth node from the end of the list and return its head.


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

from typing import Optional        
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        aNode = dummy
        bNode = head
        while n > 0:
            bNode = bNode.next
            n = n - 1
        while bNode != None:
            aNode = aNode.next
            bNode = bNode.next
        aNode.next = aNode.next.next
        return dummy.next

# 这道题可以选择用两个指针，他们之间的距离为N
# 比如说 bNode 在 aNode前两个
# 当bNode到达末尾的时候，aNode到达末尾倒数第N个
# 需要用一个dummy Node
# 需要删除倒数第N个的时候，需要在第N个前面才能删除

